Couch to SQL: missing forms, cases, and ledgers

millerdev opened the pull request:
Review by commit 🐡 

The most important change is https://github.com/dimagi/commcare-hq/pull/26768/commits/ca2b56c4e71ade51b6208602c9836dbb1390c4e5, which adds a feature to the `--forms=missing` option that attempts to migrate missing forms referenced by cases and case updates and ledger transactions that are referenced by couch but missing in SQL.

millerdev committed:
Search for missing forms in missing SQL cases

millerdev committed:
Save missing cases on migrate forms subset

millerdev committed:
Add command to filter/update diffs in state db

millerdev committed:
Process missing forms from case diffs

millerdev committed:
Classify missing couch ledger diff as "change"

millerdev committed:
Reasonable default for --missing-docs

millerdev committed:
Patch kafka to prevent migration publishing

FormProcessorSQL.save_processed_models(..., publish_to_kafka=False) was ineffective (and therefore removed) because corehq.apps.commtrack.models.publish_stock_state_to_kafka_on_save is fired when stock models are saved, regardless of the value of that flag.

millerdev commented on corehq/apps/couch_sql_migration/couchsqlmigration.py:
```diff
@@ -391,6 +394,100 @@ def _process_missing_forms(self):
+        result = self._get_case_stock_result(sql_form, couch_form)
+        if not result:
+            return False
+        cases = [c for c in result.case_models if c.has_tracked_models()]
```
This is a little scary. I'm not sure if it is always safe to save all updates ("tracked models") returned by [`self._get_case_stock_result()`](https://github.com/dimagi/commcare-hq/blob/129e08a152022dd49f6404c76a9682c4bb8a6cfa/corehq/apps/couch_sql_migration/couchsqlmigration.py#L253-L266). Will it matter if an update is applied to a case twice? Will that create a new case transaction, or will it update the existing transaction? Is there a way to tell if changes present on a given "tracked model" have already been applied (and skip them if yes)?

millerdev replied:
@snopoke I'd be interested in your feedback here.

millerdev committed:
Always classify missing forms as diffs

Previously missing forms were sometimes classified as "changes"

dannyroberts commented on corehq/apps/couch_sql_migration/couchsqlmigration.py:
```diff
@@ -595,6 +595,20 @@ def get_xml(self):
+@contextmanager
+def patch_kafka():
+    def drop_change(self, topic, change_meta):
+        log.debug("not publishing doc_id=%s to %s", change_meta.document_id, topic)
```
nit: might be useful to have the word "kafka" in here

millerdev replied:
https://github.com/dimagi/commcare-hq/pull/26768/commits/f24d48191d887eedf808c12389cef25097bf7b10

dannyroberts reviewed (approved):
I can't say I could follow every line, but overall it looks good to me. I wasn't able to weigh in on the question you asked @snopoke unfortunately, so might be worth offlining with him if that's a blocker

millerdev committed:
Add "kafka" to log message [ci skip]

snopoke commented on corehq/apps/couch_sql_migration/tests/test_migration.py:
```diff
@@ -1397,6 +1432,44 @@ def test_migrate_ledgers(self):
+        }, timedelta(days=-3))
+        form2 = self._set_balance(case_id, {self.liquorice: 75}, timedelta(days=-1))
+        print("ledger forms:", form1, form2)
+        with self.skip_case_and_ledger_updates(form1):
```
what happens if you skip `form2`, does the stock end up in the correct state with a balance of 75? I think it will not based on the current code in `_iter_missing_ledgers`

snopoke replied:
I tried this locally and was surprised to find that it didn't give any stock diffs. Digging into it more I found that it was hitting this: https://github.com/dimagi/commcare-hq/blob/082acc9bd7121a1047ef35a4c16aea0fdbd7698a/corehq/apps/couch_sql_migration/casediff.py#L193-L196

It seemed odd that there would be duplicate transaction but digging deeper I found a bug in `has_duplicate_transactions`. This method compares the number of DB transactions with the number in the form. (get read to cringe?) This doesn't work because the legacy backend creates an additional transaction in certain cenarios:

https://github.com/dimagi/commcare-hq/blob/579b52f1037d30c8e9184214f7278bb1d6a8b9d2/corehq/ex-submodules/casexml/apps/stock/signals.py#L28-L47

With this change to `StockTransactionLoader.iter_stock_transactions` I got the diffs I was expecting:

```diff
     def iter_stock_transactions(self, form_id):
         xform = FormAccessorCouch.get_form(form_id)
         assert xform.domain == _diff_state.domain, xform
         for report in get_all_stock_report_helpers_from_form(xform):
             for tx in report.transactions:
                 yield report.report_type, tx
+                if tx.action == TRANSACTION_TYPE_STOCKONHAND:
+                    yield report.report_type, tx
```

(I think this change is all that is necessary since they are only used to count the number of transactions).

millerdev replied:
@snopoke I think the concerns you raise are resolved by https://github.com/dimagi/commcare-hq/pull/26768/commits/fc221210fd95a8b05a5d8a03cbe3e6310d3093cf Please verify that the test I added matches what you meant by "what happens if you skip `form2`?"

snopoke replied:
yes, that resolves it. I think the rebuild was the missing piece that was causing diffs even after fixing the comparison.

millerdev commented:
Thanks for the very detailed feedback, @snopoke!

millerdev committed:
Count changed ledgers

millerdev committed:
Fix missing ledger diffs

- Add extra STOCKONHAND transaction to match behavior of legacy backend.
- Rebuild ledger before saving missing ledger transactions.

millerdev commented on corehq/apps/couch_sql_migration/couchsqlmigration.py:
```diff
@@ -391,6 +394,100 @@ def _process_missing_forms(self):
+        return migrated_ids
+
+    def _save_missing_cases_and_ledgers(self, form_id):
+        def iter_missing_ledgers(stock_result):
```
> If there is a concern that the `LedgerValue` might exist but may be missing some transactions

Yes, this is the situation I'm trying to resolve here.

> then you should use `transaction.natural_key()` to look for missing ones (not `transaction.ledger_reference`) since the natural key includes `form_id`.

The result of `LedgerAccessorSQL.get_ledger_transactions_for_form(form_id, case_ids)` only includes transactions relevant to *one form*, and the ledger values I'm checking all came out of that form. It should yield the same result whether this code matches on `ledger_reference` or `natural_key()`, right?

I am curious if it is possible for a single form to have more than one ledger transaction for a given ledger value? I think this is not supported, based on the data model keys, but I may have overlooked something.

> If it turns out that a transaction is missing for an existing `LedgerValue` then you'd also need to rebuild the `LedgerValue`.

I think this is at least part of what is missing. Hopefully it will resolve the issue outlined in https://github.com/dimagi/commcare-hq/pull/26768/files/082acc9bd7121a1047ef35a4c16aea0fdbd7698a#r388588292

snopoke replied:
> It should yield the same result whether this code matches on ledger_reference or natural_key(), right?

True, in this case it would since all the form_id values would be the same

>I am curious if it is possible for a single form to have more than one ledger transaction for a given ledger value? I think this is not supported, based on the data model keys, but I may have overlooked something.

Interestingly you can. There's nothing to prevent that in the code. I don't think that's an intentional decision though. Would be interesting to see if that case exists in the wild.
